### ISSUE TITLE
Any idea how to add billing to your app without GraphQL

### ISSUE THREAD

**Issue by Kurbm** [kirillplatonov/shopify-hotwire-sample#56]
Hey,

does anyone has an idea, how to add billing with the updated template?

The way I did it with a full-page redirect, does not work anymore.

Thanks and best,
Kurbm

**Comment by retgoat** [comment]
Having the same issue. 
The app renders redirect view, but rails says that form submission needs to redirect user afterwards. 
If I disable turbo with `data: { turbo: false }` then shop auth does not work:
```
Redirected to https://4cca-87-116-167-77.ngrok.io/?host=cXRlZWR5LWRldi5teXNob3BpZnkuY29tL2FkbWlu&return_to=%2Fsubscriptions%3Fshop%3Dqteedy-dev.myshopify.com&shop=qteedy-dev.myshopify.com
10:03:44 web.1     | Filter chain halted as :redirect_to_splash_page rendered or redirected
10:03:44 web.1     | Completed 302 Found in 1ms (ActiveRecord: 0.0ms | Allocations: 259)
```

@kirillplatonov any ideas how subscription could be made with Shopify REST API?
Thanks in advance!

**Comment by retgoat** [comment]
Hm, it seems you were my rubber ducks :)
Here is the approach to create a subscription with REST API:

```ruby
# subscription_controller.rb
def create
  ShopifyAPI::Auth::Session.temp(shop: current_shop.shopify_domain, access_token: current_shop.shopify_token) do |session|
    recurring_charge = ShopifyAPI::RecurringApplicationCharge.new(session: session)
    recurring_charge.name = name
    recurring_charge.price = price
    recurring_charge.test = !Rails.env.production?
    recurring_charge.return_url = subscription_confirm_url(@plan.id, host: ENV['HOST'], protocol: "https")

    if recurring_charge.save!
      turbo_redirect_to confirmation_redirect_path(url: recurring_charge.confirmation_url)
    else
      flash[:alert] = recurring_charge.errors.messages.join("\n")
      turbo_redirect_to subscriptions_index_path
    end
  end
end

def confirmation_redirect

end


# confirmation_redirect.html.erb

<%= polaris_page(title: "Confirm subscription") do |page| %>
  <%= polaris_callout_card(
    title: "Confirm subscription",
    illustration: "https://cdn.shopify.com/s/assets/admin/checkout/settings-customizecart-705f57c725ac05be5a34ec20c05b94298cb8afd10aac7bd9c7ad02030f48cfa0.svg",
  ) do |callout| %>
    <div data-controller="confirmation-redirect">
      <div id="redirect-url" data-url="<%= params[:url] %>"></div>
      <% callout.primary_action(url: params[:url]) { "Confirm Subscrition" } %>

      <p>In order to complete subscription we need to redirect you. Please complete the subscription by hitting the button below</p>
      <p>You will be automatically redirected after <span id="seconds-left">5</span> seconds</p>
    </div>
  <% end %>
<% end %>
```
```js
// confirmation_redirect_controller.js

import { Controller } from "@hotwired/stimulus"

// Connects to data-controller="confirmation-redirect"
export default class extends Controller {
  connect() {

    const url = document.getElementById("redirect-url").dataset['url']

    var el = document.getElementById('seconds-left'),
      total = el.innerHTML,
      timeinterval = setInterval(function () {
        total = --total;
        el.textContent = total;
        if (total <= 0) {
          clearInterval(timeinterval);
          window.top.location = url
        }
      }, 1000);
  }
}
```

After 5 seconds user gets redirected to subscription approval. After approve I successfully got them back into app to finish subscription setup

Hope that helps

**Comment by EmersonEmerson** [comment]
[This may be unrelated to method desired in question - i was a bit confused on how to get billing working from the sample]

I was able to accomplish billing w/ low lift by doing the following:
1. You need a fork of shopify_app with[ these changes](https://github.com/Shopify/shopify_app/pull/1578/commits/e002eda9a09b6b836607784b2e9257b5bb595255) to fix the infinite redirect issue. [(From @kirillplatonov 's outstanding pull req here.)](https://github.com/Shopify/shopify_app/pull/1578)
```
# Gemfile

gem "shopify_app", git: 'https://github.com/REPLACE/shopify_app'

```

2. You can simply use the billing config [as shown here.](https://github.com/kirillplatonov/shopify-hotwire-sample/compare/main...billing#diff-f5387ade26cb7fddeb700ce50ac404c7480ba27bbe5342160653feb10b0dadee)

**Comment by EmersonEmerson** [comment]
Hey again,

So when using the billing.config solution in initializers/shopify_app.rb... I wasn't sure how to do all the cool stuff I could do before, like:
* Allow dev stores to bypass the charge
* Don't charge certain users
* Multiple plans or tiers

(If it is possible to only fire the billing.config for certain stores - im all ears... seems all or nothing but I am a neanderthal)


**As such, I've been digging into this and think I have a pretty simple (albeit hack-y) solution...**

In short, from what you have currently @retgoat:
1. Remove the redirects at the end of subscription_controller#create
2. Add a create.html.erb in app/views/subscription/ 
3. Back in subscription_controller#create, pass along the confirmation URL to the view.
4. In the view file:
`<%= javascript_tag do %>
window.top.location.href = "<%= @recurring_application_charge.confirmation_url %>";
<% end %>`

**Comment by EmersonEmerson** [comment]
Hey again,

So when using the billing.config solution in initializers/shopify_app.rb... I wasn't sure how to do all the cool stuff I could do before, like:
* Allow dev stores to bypass the charge
* Don't charge certain users
* Multiple plans or tiers

If it is possible to only fire the billing.config for certain stores - im all ears... seems all or nothing but I am a neanderthal... In the meantime I've switched back to the rest API.

In the past, I had fixed redirect issue by using a view which fires JS to redirect like so:
`<%= javascript_tag do %>
window.top.location.href = "<%= @recurring_application_charge.confirmation_url %>";
<% end %>`

Your solution looks much more elegant @retgoat , appreciate the share!
